### DISCUSSION TITLE
Preferred Error Handling

### DISCUSSION THREAD

**Discussion by hecon5** [joyfullservice/msaccess-vcs-addin#438] [General]
I've noticed `DebugMode` and `LogUnhandledErrors` are now in their own module, which is good I think. I also noticed that newer functions seem to use `LogUnhandledErrors` instead of `DebugMode`, is `LogUnhandledErrors` the new preferred mechanism for handling Errors? From reading how `DebugMode` seems to work, it simply points to `LogUnhandledErrors`, but thought I'd check and see if I'm missing something.

**Comment by joyfullservice** [comment]
Great question! Yes, your observation is correct. My original approach was to use the `DebugMode` function as part of an inline error handling switch. It was concise and effective, but had one main problem. `DebugMode` reads and acts like a property, but it also contained an important side effect that was not readily apparent on the surface. 

Because an `On Error` directive silently discards any existing error condition, otherwise untrapped errors would be lost in the process. By using this function as part of my conditional `On Error` directive, I was able to effectively log or break on these errors. I suppose I could have been more descriptive with the function name, but I wasn't excited about `GetDebugModeAfterCheckingForUnhandledErrors`....  😉 It also doesn't do so great with the single responsibility principle. I later added a required parameter to give a bit of a clue about something else taking place, but I still wasn't thrilled with it.

While *I* knew what `DebugMode` meant, it was understandably confusing for new developers on the project. I am sure they were puzzling over the sanity of the following line:

```vba
If DebugMode(True) Then On Error Resume Next Else On Error Resume Next`
```

After a lot of thought and consideration, I finally settled on making this two lines with hopefully a greater level of clarity on what is actually taking place. [Less clever](https://nolongerset.com/dont-write-clever-code/) and more clear.

```vba
' Old way to set On Error Resume Next
If DebugMode(True) Then On Error Resume Next Else On Error Resume Next

' New (preferred) way
LogUnhandledErrors
On Error Resume Next

' Enabling inline debugging remains the same
If DebugMode(True) Then On Error GoTo 0 Else On Error Resume Next
```

Inline debugging is mainly intended for debugging functionality in the add-in where it is useful to jump right to the line causing the error. As we understand more about the specific errors being thrown, we can use the `Catch` functions to appropriately deal with the specific errors.
